### PR TITLE
[NVPTX] Lower 16xi8 and 8xi8 stores efficiently

### PR DESCRIPTION
**Pull request by bondhugula**
Lower 16xi8 vector stores in NVPTX ISel efficiently using
st.v4.b32 instead of multiple st.v4.u8 along the lines of vector loads
and 8xf16. Similarly, 8xi8 using st.v2.u32.

### CODE REVIEW

File: llvm/test/CodeGen/NVPTX/vector-stores.ll [hunk 1]
```diff
@@ -37,3 +37,17 @@ define void @v16i8(ptr %a, ptr %b) {
   store <16 x i8> %v, ptr %b
   ret void
 }
+
+; CHECK-LABEL: .visible .func v16i8_store
+define void @v16i8_store(ptr %a, <16 x i8> %v) {
+; CHECK: st.v4.u32
```
Comment (ldrumm) [inline llvm/test/CodeGen/NVPTX/vector-stores.ll:43]: Can we check for the extracts as well (their absense)?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 2]
```diff
@@ -5557,6 +5557,50 @@ static SDValue PerformLOADCombine(SDNode *N,
       DL);
 }
 
+// Lower a v16i8 (or a v8i8) store into a StoreV4 operation with i32 results
+// instead of letting ReplaceLoadVector split it into smaller stores during
+// legalization. This is done at dag-combine1 time, so that vector operations
```
Comment (ldrumm) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5562]:
```suggestion
// legalization. This is done at dag-combine time, so that vector operations
```

**Review by ldrumm — APPROVED** [review]
Minor nits. LGTM

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 2 above)
Comment (bondhugula) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5562]: Are you sure about this? This is intended to be done before legalization. Changed to dag-combine.

File: llvm/test/CodeGen/NVPTX/vector-stores.ll (hunk 1 above)
Comment (bondhugula) [reply llvm/test/CodeGen/NVPTX/vector-stores.ll:43]: Sure, done.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 2 above)
Comment (ldrumm) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: My mistake; I read it as a typo

**Review by Artem-B — APPROVED** [review]

**Comment by steven-johnson** [comment]
This seems to have injected failures into Halide codegen; we are now getting runtime errors of the form `CUDA_ERROR_MISALIGNED_ADDRESS` for `cuMemcpyDtoH()` where we didn't before. It appears we are now emitting an aligned store instruction where we previous emitted an unaligned one. Can we get a revert of this pending further investigation, please?

File: llvm/test/CodeGen/NVPTX/vector-stores.ll [hunk 3]
```diff
@@ -37,3 +37,19 @@ define void @v16i8(ptr %a, ptr %b) {
...
+  ; CHECK-NEXT: ld.param.v4.u32   {%r1, %r2, %r3, %r4}, [v16i8_store_param_1];
+  ; CHECK-NEXT: st.v4.u32   [%rd1], {%r1, %r2, %r3, %r4};
+  store <16 x i8> %v, ptr %a
+  ret void
+}
+
+; CHECK-LABEL: .visible .func v8i8_store
+define void @v8i8_store(ptr %a, <8 x i8> %v) {
+  ; CHECK: st.v2.u32
+  store <8 x i8> %v, ptr %a
```
Comment (steven-johnson) [inline llvm/test/CodeGen/NVPTX/vector-stores.ll:53]: This is only correct if the pointer is aligned to a 4-byte-boundary (IIUC), but AFAIK nothing in the IR to this point promises that alignment

Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/vector-stores.ll:53]: You're right. Using larger types for loads/stores must be aligned appropriately.

We do use `allowsMemoryAccessForAlignment` in other places.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 4]
```diff
@@ -5557,6 +5557,51 @@ static SDValue PerformLOADCombine(SDNode *N,
       DL);
 }
 
+// Lower a v16i8 (or a v8i8) store into a StoreV4 (or StoreV2) operation with
+// i32 results instead of letting ReplaceLoadVector split it into smaller stores
+// during legalization. This is done at dag-combine time, so that vector
+// operations with i8 elements can be optimised away instead of being needlessly
+// split during legalization, which involves storing to the stack and loading it
```
[the remark this answers is not in the document]
Comment (pasaulais) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5564]: Note that this comment might be out of date, as it looks copied from `PerformLOADCombine` and that was written before stack optimizations were done

**Review by pasaulais — COMMENTED** [review]
LGTM once the alignment issue is addressed

**Pull request merged**